src/native/minipal/thread.h

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifndef HAVE_MINIPAL_THREAD_H
+#define HAVE_MINIPAL_THREAD_H
+
+#include <pthread.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdlib.h>
+
+#if defined(__linux__)
+#include <unistd.h>
+#include <sys/syscall.h>
+#elif defined(__FreeBSD__)
+#include <pthread_np.h>
+#elif defined(__NetBSD__)
+#include <lwp.h>
+#elif defined(__HAIKU__)
+#include <kernel/OS.h>
+#endif
+
+#ifdef PTHREAD_MAX_NAMELEN_NP
+#define MINIPAL_MAX_THREAD_NAME_LENGTH (PTHREAD_MAX_NAMELEN_NP - 1)
+#elif defined(__APPLE__)
+#define MINIPAL_MAX_THREAD_NAME_LENGTH 63
+#elif defined(__FreeBSD__)
+#define MINIPAL_MAX_THREAD_NAME_LENGTH MAXCOMLEN
+#elif defined(__HAIKU__)
+#define MINIPAL_MAX_THREAD_NAME_LENGTH (B_OS_NAME_LENGTH - 1)
+#else
+#define MINIPAL_MAX_THREAD_NAME_LENGTH 15
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Get the current thread ID.
+ *
+ * @return The current thread ID as a size_t value.
+ */
+static inline size_t minipal_get_current_thread_id(void)
+{
+#ifdef __wasm
+    return 0;
+#else
+#if defined(__GNUC__) && !defined(__clang__) && defined(__cplusplus)
+    // gcc doesn't like _Thread_local when __cplusplus is defined.
+    // although thread_local is C2x, which other compilers don't allow with C11.
+    static thread_local size_t tid = 0;
```

Comment:
It was this hack for gcc, or using a dedicated .c file. The latter was giving duplicate symbol errors on macCatalyst
```
duplicate symbol '_minipal_get_current_thread_id' in:
    /Users/adeel/projects/runtime5/artifacts/bundles/tests/System.Runtime.Tests/libSystem.Native.a[23](thread.c.o)
    /Users/adeel/projects/runtime5/artifacts/bundles/tests/System.Runtime.Tests/libmonosgen-2.0.a[70](thread.c.o)
duplicate symbol '_minipal_set_thread_name' in:
    /Users/adeel/projects/runtime5/artifacts/bundles/tests/System.Runtime.Tests/libSystem.Native.a[23](thread.c.o)
    /Users/adeel/projects/runtime5/artifacts/bundles/tests/System.Runtime.Tests/libmonosgen-2.0.a[70](thread.c.o)
ld: 2 duplicate symbols
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
If necessary, it will be possible to split into .c file later on once https://github.com/dotnet/runtime/pull/107889 is merged.

Reply:
`thread_local` is supposed to be a macro in `<threads.h>` in C11 (changed to a keyword in C23).
No idea how it would behave in C++ mode, though.

I don't have a strong opinion about the hack vs just using __thread.

Reply:
Yup, I think it can be simplified as part of https://github.com/dotnet/runtime/pull/107889 or after that one is merged (I'll do a round on getexepath and this). For now, it's a minor spaghetti we can live with ig.